### ISSUE TITLE
Database source archives in the workspace slow down VS Code search

### ISSUE THREAD

**Issue by adityasharad** [github/vscode-codeql#1068]
**Describe the bug**
Multiple internal users report that:
- having at least one database open in the workspace makes VS Code's built-in search unusably slow, and
- removing the database or its source archive from the workspace fixes the search speed.

**Expected behavior**
Users should be able to add as many CodeQL databases as they wish to their workspace without affecting the search function.

**Additional context**
Related: https://github.com/github/vscode-codeql/issues/1065

--- 
Tentative diagnosis: we add database source archive folders as workspace folders, which means VS Code's search will attempt to search within these folders. These may contain a large number of files, degrading search performance.

Possible solution: When adding a source archive folder to the workspace, modify the workspace folder settings for that specific folder, and exclude it from VS Code search. See the `update` method in [WorkspaceConfiguration](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration).
This will mean users cannot search within the source archive (except by running a query or using the AST viewer), but arguably that is not a feature we want to support anyway.

**Comment by hmac** [comment]
I did a little bit of digging into this, out of curiosity.

Firstly, I can reproduce the problem even with very small databases - e.g. a Ruby database generated from a single empty `foo.rb` file, which is 4 MB and contains 51 files. So my hunch is that it isn't caused by vscode having to search through a lot of files.

When I have a database added and I run a search, I noticed the vscode console emits this warning:
> `No search provider registered for scheme: codeql-zip-archive, waiting`

That message is logged [here](https://cs.github.com/microsoft/vscode/blob/3e70603d63d3dca48f7d587a4118521af7be8013/src/vs/workbench/services/search/common/searchService.ts#L199), where it looks like vscode dispatches the file search to separate providers for each scheme (`file` is the default scheme, and I guess we have a custom scheme for databases called `codeql-zip-archive`). It won't return results until all providers give back a response, even if there's no provider registered for a scheme (I think - and this seems like a bug that should probably be fixed upstream).

Manually inserting a dummy provider in the context of that function with
```js
this.registerSearchResultProvider("codeql-zip-archive", 0, {
  textSearch: function(q, o, t) { return null; },
  fileSearch: function(q, t) { return null; },
  clearCache: function(s) { return null; }
})
```
seems to fix the issue - I get instant results for files in my normal workspace.

Obviously this means I can't search inside any databases, but I don't tend to do that anyway.

I hope this helps!

**Comment by adityasharad** [comment]
Thank you, that looks like the problem! Looks like VS Code assumes all schemes have a provider, and a missing provider is just being slow to initialise rather than nonexistent.

For reference -- the Search Provider API doesn't seem to be in the API docs yet, but it's mentioned here: https://code.visualstudio.com/updates/v1_24#_search-provider-api

**Comment by aeisenberg** [comment]
Short term, we should add a no-op search provider, but I wonder if there is a way to actually provide text search in source archives. This limitation has come up before.

**Comment by adityasharad** [comment]
Update from investigation with @shati-patel: search providers are currently a [proposed API](https://code.visualstudio.com/api/advanced-topics/using-proposed-api), so we can use them in local debug builds but not in a published extension release.

**Comment by shati-patel** [comment]
Upstream issue opened here: https://github.com/microsoft/vscode/issues/140737

Two things I didn't fully investigate in the end:
1. The "Ctrl+Shift+F" search worked fine for me, so I didn't dig around there too much, but mentioned it briefly anyway. If this turns out to be an independent problem, I can open a separate issue 🙂 
2. I spent a while trying to get the proposed API to work, but it wasn't compatible with the VS Code API we currently use... I _started_ updating dependencies and fixing errors, but it was turning into more work than it was probably worth (giving that we can't officially use it anyway!) 

We'll see what the VS Code folks come up with 🤞🏽 ✨

**Comment by shati-patel** [comment]
The VS Code team have pushed a fix (https://github.com/microsoft/vscode/commit/72aaaef92be8317a9fb40c4997bd5ef9f8c9f3f1), which is live for VS Code Insiders. I'll keep this issue open until it's also part of the main release 🙂

**Comment by aeisenberg** [comment]
That's great. Have you confirmed this is fixed for codeql on the insiders edition?

**Comment by shati-patel** [comment]
> That's great. Have you confirmed this is fixed for codeql on the insiders edition?

Yep! All good with Insiders 😌

**Comment by aeisenberg** [comment]
Thanks for keeping track of this. I think we should keep this issue open for now, until the fix makes it into a proper release of vscode. This way, if anyone else is seeing the same problem, they will know that it is fixed in insiders.

**Comment by shati-patel** [comment]
Fixed in VS Code 1.64.0!
